### ISSUE TITLE
JRuby+Truffle: Keyword Arguments - "illegal recursive call"

### ISSUE THREAD

**Issue by fniephaus** [jruby/jruby#2541]
Hi folks,

If you run the following code with JRuby+Truffle:

```ruby
def m(i:)
    i
end

50000.times.each do
  (1..100_000).each do |i|
    m i:i
  end
end
```

you'll get something similar to this:

```
jruby: using Graal Truffle Runtime
[truffle] opt fail  block in each:./core/kabench.rb:6 <split-0-U> |Reason com.oracle.graal.nodes.util.GraphUtil$2: Found illegal recursive call to HotSpotMethod<Class.getCanonicalName()>, must annotate such calls with @TruffleBoundary! 
[truffle] opt fail  m:./core/kabench.rb:1 <split-0-U>             |Reason com.oracle.graal.nodes.util.GraphUtil$2: This code path should never be part of a compilation. [] 
[truffle] opt fail  Range#each(core):core: Range#each <split-0-U> |Reason com.oracle.graal.nodes.util.GraphUtil$2: This code path should never be part of a compilation. [] 
```

I assume that there's just a TruffleBoundary missing somewhere as suggested by the error message. But unfortunately, I'm unable to tell if that's really the case and if yes how to fix it.
Any ideas?

Regards,
Fabio

**Comment by chrisseaton** [comment]
Our current implementation of keyword arguments doesn't compile - it only works in the interpreter. You're supposed to get a nice error message telling you this, due to the `notDesignedForCompilation` directives (e.g. https://github.com/jruby/jruby/blob/1095f6785e2b6f8f884b25d40e2074b52893667b/truffle/src/main/java/org/jruby/truffle/nodes/methods/arguments/ReadKeywordArgumentNode.java#L36) but for some reason the compiler here is failing before it gets a chance to report that error. That happens sometimes.

If you change `m` to a keyword-rest argument, as in `m(**i)`, you do get the nice compiler error.

The reason the keyword arguments don't compile is that we've taken shortcuts and used slow operations such as `verySlowToKeyValues` and Java's `equals` rather than setting up proper Ruby call sites yet.

**Comment by fniephaus** [comment]
Closed, see #2567.
